### DISCUSSION TITLE
How do I run extra num_iteration with a LightGBMTuner Study?

### DISCUSSION THREAD

**Discussion by wil70** [optuna/optuna#5273] [Q&A]
### Don't use GitHub Issues to ask support questions.

Hello, 

**1) How do I run more iteration of a LightGBMTuner Study?**
I'm attempting to perform an additional 900 iterations after the initial 100, but I'm encountering difficulties. I adapted an example from Optuna + LightGBM, replacing the training with LightGBMTuner. Here is the code:

```
import optuna
import optuna.integration.lightgbm as lgb
import numpy as np
from lightgbm import early_stopping
from lightgbm import log_evaluation
import sklearn.datasets
from sklearn.model_selection import KFold
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score

if __name__ == "__main__":
    data, target = sklearn.datasets.load_breast_cancer(return_X_y=True)
    #dtrain = lgb.Dataset(data, label=target)
    #train_x, valid_x, train_y, valid_y = train_test_split(data, target, test_size=0.25)
    train_x, test_x, train_y, test_y = train_test_split(data, target, test_size=0.10)
    train_x, valid_x, train_y, valid_y = train_test_split(train_x, train_y, test_size=0.25)
    dtrain = lgb.Dataset(train_x, label=train_y)
    dval = lgb.Dataset(valid_x, label=valid_y)

    params = {
        "objective": "binary",
        "metric": "binary_error",
        "verbosity": -1,
        "boosting_type": "dart",
    }

    modelDir="QuickTestTunerModel"

    try:
        study = optuna.load_study(study_name='study-wil10_8_v2', storage='sqlite:///'+modelDir+'\\wil10_8_v2.db')
    except KeyError:
        study = optuna.create_study(study_name='study-wil10_8_v2', storage='sqlite:///'+modelDir+'\\wil10_8_v2.db', direction="minimize")

    tuner = lgb.LightGBMTuner(
            params,
            dtrain,
            valid_sets=[dval], #####
            study=study,
            num_boost_round=100,
            #folds=KFold(n_splits=3),
            callbacks=[early_stopping(100), log_evaluation(100)],
            model_dir=modelDir,
            keep_training_booster =True,
            #return_cvbooster=True
    )
    

    tuner.run()
    tuner.run()
    
    # Run an additional 100 trials
    #study.optimize(n_trials=100)
    #tuner.run()

    print("Best score:", tuner.best_score)
    print("Best booster score:", tuner._get_booster_best_score)
    best_params = tuner.best_params
    print("Best params:", best_params)
    print("  Params: ")
    for key, value in best_params.items():
        print("    {}: {}".format(key, value))

 
    # Save the best model
    tuner.get_best_booster().save_model("QuickTestTunerModel\\bestModel.txt")

    # Load the saved model
    bst = tuner.get_best_booster() #lgb.Booster(model_file='bestModel.txt')

    preds = bst.predict(train_x)
    preds = np.rint(preds)
    accuracy = accuracy_score(train_y, preds)
    print("train accuracy: "+str(accuracy))
    
    preds = bst.predict(valid_x)
    preds = np.rint(preds)
    accuracy = accuracy_score(valid_y, preds)
    print("valid accuracy: "+str(accuracy))

    # Predict on the testing data
    preds = bst.predict(test_x)
    preds = np.rint(preds)
    accuracy = accuracy_score(test_y, preds)
    print("test accuracy: "+str(accuracy))
```

The output look like this

```
Best score: 0.0234375
Best booster score: <bound method _BaseTuner._get_booster_best_score of <optuna.integration._lightgbm_tuner.optimize.LightGBMTuner object at 0x000001FAAAEA5460>>
Best params: {'objective': 'binary', 'metric': 'binary_error', 'verbosity': -1, 'boosting_type': 'dart', 'feature_pre_filter': False, 'lambda_l1': 0.0, 'lambda_l2': 0.0, 'num_leaves': 31, 'feature_fraction': 0.8, 'bagging_fraction': 1.0, 'bagging_freq': 0, 'min_child_samples': 20}
  Params:
    objective: binary
    metric: binary_error
    verbosity: -1
    boosting_type: dart
    feature_pre_filter: False
    lambda_l1: 0.0
    lambda_l2: 0.0
    num_leaves: 31
    feature_fraction: 0.8
    bagging_fraction: 1.0
    bagging_freq: 0
    min_child_samples: 20
train accuracy: 1.0
valid accuracy: 0.9765625
test accuracy: 0.9824561403508771
```
The second time I run the program it load from the sqllite database but it doesn't run more iterations. I tried adding an extra tuner.run() and other things like 'study.optimize(n_trials=100)' but I couldn't make it work.

Output of the second run:
```
 0%|                                                                                                                                                                                                                                                                                                                                                                                                                        | 0/7 [00:00<?, ?it/s]
  0%|                                                                                                                                                                                                                                                                                                                                                                                                                       | 0/20 [00:00<?, ?it/s] 
  0%|                                                                                                                                                                                                                                                                                                                                                                                                                       | 0/10 [00:00<?, ?it/s] 
  0%|                                                                                                                                                                                                                                                                                                                                                                                                                        | 0/6 [00:00<?, ?it/s]
  0%|                                                                                                                                                                                                                                                                                                                                                                                                                       | 0/20 [00:00<?, ?it/s] 
  0%|                                                                                                                                                                                                                                                                                                                                                                                                                        | 0/5 [00:00<?, ?it/s] 
  0%|                                                                                                                                                                                                                                                                                                                                                                                                                        | 0/7 [00:00<?, ?it/s] 
  0%|                                                                                                                                                                                                                                                                                                                                                                                                                       | 0/20 [00:00<?, ?it/s] 
  0%|                                                                                                                                                                                                                                                                                                                                                                                                                       | 0/10 [00:00<?, ?it/s] 
  0%|                                                                                                                                                                                                                                                                                                                                                                                                                        | 0/6 [00:00<?, ?it/s]
  0%|                                                                                                                                                                                                                                                                                                                                                                                                                       | 0/20 [00:00<?, ?it/s] 
  0%|                                                                                                                                                                                                                                                                                                                                                                                                                        | 0/5 [00:00<?, ?it/s] 
Best score: 0.0234375
Best booster score: <bound method _BaseTuner._get_booster_best_score of <optuna.integration._lightgbm_tuner.optimize.LightGBMTuner object at 0x000002329EA68410>>
Best params: {'objective': 'binary', 'metric': 'binary_error', 'verbosity': -1, 'boosting_type': 'dart', 'feature_pre_filter': False, 'lambda_l1': 0.0, 'lambda_l2': 0.0, 'num_leaves': 31, 'feature_fraction': 0.8, 'bagging_fraction': 1.0, 'bagging_freq': 0, 'min_child_samples': 20}
  Params:
    objective: binary
    metric: binary_error
    verbosity: -1
    boosting_type: dart
    feature_pre_filter: False
    lambda_l1: 0.0
    lambda_l2: 0.0
    num_leaves: 31
    feature_fraction: 0.8
    bagging_fraction: 1.0
    bagging_freq: 0
    min_child_samples: 20
train accuracy: 0.9895833333333334
valid accuracy: 1.0
test accuracy: 1.0

```

Despite these attempts, the program doesn't run more iterations. How can I achieve this?

**2) Potential Issues:**
Upon the second run, I observed differences in training accuracy. Initially at 1, it decreased later. Meanwhile, validation and test accuracy increased to 1. This behavior seems unusual. Any insights into this?

Thank you for your assistance.

--Wil

**Comment by wil70** [comment]
For (1) I would expect something like "tuner.run()" or "study.optimize(n_trials=100)"  to do 100 more iterations

(2) look like a bug, isn't it?

**Comment by nzw0301** [comment]
Could you clarify the difference between `iteration` and `n_trials` in your question because there is no `900` in the code? I'm not sure what is the question.

> (2)

I don't think so because the shared script does not fix random seed values even if each optimisation is not dependent on the others.

**Comment by wil70** [comment]
Thanks @nzw0301 

1) I ran this code with 100 epochs max (num_boost_round) and then it tries new parameters until it is done. Then I saved the study. Now I would like to run an extra 100 'study iterations' so it doesn't compute again with the optuna parameters it aready tried earlier. Like if I did num_boost_round=200 (100+100) from the start instead of the num_boost_round=100 as I did. 
I do not see how to do this with optuna **lgb.LightGBMTuner()** from a saved sutdy?

(I think sometime you see epochs, [num_iterations](https://lightgbm.readthedocs.io/en/latest/Parameters.html#num_iterations), num_boost_round, n_trials, n_iter, num_tree, num_trees, num_round, num_rounds, nrounds, num_boost_round, n_estimators, max_iter - this can be confusing)

I also tried to call tuner.run() hopping it will do another round of 100, I also tried to change num_boost_round=from 100 to 200, but that didn't trigger some extra computation from a loaded study with LightGBMTuner

```
0%|                                                                                                                                                                                                                                                                                 | 0/7 [00:00<?, ?it/s]
  0%|                                                                                                                                                                                                                                                                                | 0/20 [00:00<?, ?it/s] 
  0%|                                                                                                                                                                                                                                                                                | 0/10 [00:00<?, ?it/s] 
  0%|                                                                                                                                                                                                                                                                                 | 0/3 [00:00<?, ?it/s] 
  0%|                                                                                                                                                                                                                                                                                | 0/20 [00:00<?, ?it/s] 
  0%|                                                                                                                                                                                                                                                                                 | 0/5 [00:00<?, ?it/s] 
Best score: 0.0234375
```


2) potential bug?

I added the following parameters to the parmas "input_model": "QuickTestTunerModel\\bestModel.txt"
and I commented out the "  #tuner.get_best_booster().save_model("QuickTestTunerModel\\bestModel.txt")"
and now the result stays the same as I expected. 

I do not understand why the save_model change the prediction accuracy result (with no retraining) from 
  train accuracy: 1.0
  valid accuracy: 0.9765625
  test accuracy: 0.9824561403508771
To
  train accuracy: 0.9895833333333334
  valid accuracy: 1.0
  test accuracy: 1.0

Later I remove the comment for "save_model" and compared the saved models after several run, and they are the same

To recap: I'm not understanding why those accuracies are different if "save_model" is called and knowing those saved model file are identical. If save_model is commented, then result are the same as expected.

Note: changed title to try to make it clearer
